pdi-execution-engine/api/src/main/java/org/pentaho/di/engine/api/IOperation.java

```diff
@@ -0,0 +1,31 @@
+package org.pentaho.di.engine.api;
+
+import java.util.List;
+
+/**
+ * IOperation roughly corresponds to a Step in PDI.
+ * An IOperation represents the logical structure of
+ * a step within the trans graph.  Materialization of
+ * an IOperation converts it to an ICallableOperation,
+ * which is associated with the behavior specific to
+ * an IEngine.
+ */
+public interface IOperation {
+
+  String getId();
+
+  List<IOperation> getFrom();
+  List<IOperation> getTo();
+
+  String getConfig();
+
+  // how to represent and manage config metadata?
+
+  /**
+   * Accepts an IOperationVisitor.  Used for inspecting or modifying
+   * the Trans graph.
+   */
+  <T> T accept( IOperationVisitor<T> visitor );
```

Comment:
My thought with IOperationVisitors is (I think) essentially what you were envisioning with a MaterializationVisitor. IOperationVisitors would walk the full operation graph, from source to sink, potentially extracting some information about the structure, and even producing a reordered or entirely new graph.  IOperationVisitors could be registered with the engine as you describe with MaterializationVisitors.

What nodes would .accept() an IMaterializationVisitor?